File: StripMate/Models/Strip.swift
import Foundation
import SwiftData

@Model
public final class Strip {
    @Attribute(.unique) public var id: String
    public var senderId: String
    private var receiverIdsString: String = ""
    public var imageUrl: String
    public var timestamp: Date
    public var latitude: Double?
    public var longitude: Double?
    public var cityName: String?
    public var thumbnailUrl: String?
    public var smallThumbnailUrl: String?
    public var flagged: Bool
    public var flagReason: String?
    public var voiceUrl: String?
    public var isSecret: Bool = false
    private var unlockedByString: String = ""
    private var seenByString: String = ""

    public var unlockedBy: [String] {
        get { unlockedByString.isEmpty ? [] : unlockedByString.split(separator: ",").map(String.init) }
        set { unlockedByString = newValue.joined(separator: ",") }
    }

    public var seenBy: [String] {
        get { seenByString.isEmpty ? [] : seenByString.split(separator: ",").map(String.init) }
        set { seenByString = newValue.joined(separator: ",") }
    }

    /// Bu strip gizli mi ve henüz userId tarafından açılmamış mı?
    public func isLockedFor(_ userId: String) -> Bool {
        guard isSecret, senderId != userId else { return false }
        return !unlockedBy.contains(userId)
    }

    public var receiverIds: [String] {
        get {
            receiverIdsString.split(separator: ",").map(String.init)
        }
        set {
            receiverIdsString = newValue.joined(separator: ",")
        }
    }
    
    // Convert to older struct when needed for non-SwiftData components
    public var asMetadata: PhotoMetadata {
        PhotoMetadata(
            id: id,
            senderId: senderId,
            receiverIds: receiverIds,
            imageUrl: imageUrl,
            timestamp: timestamp,
            latitude: latitude,
            longitude: longitude,
            cityName: cityName,
            thumbnailUrl: thumbnailUrl,
            smallThumbnailUrl: smallThumbnailUrl,
            flagged: flagged,
            flagReason: flagReason,
            voiceUrl: voiceUrl,
            isSecret: isSecret,
            unlockedBy: unlockedBy,
            seenBy: seenBy
        )
    }
    
    public init(id: String = UUID().uuidString, senderId: String, receiverIds: [String] = [], imageUrl: String, timestamp: Date = Date(), latitude: Double? = nil, longitude: Double? = nil, cityName: String? = nil, thumbnailUrl: String? = nil, smallThumbnailUrl: String? = nil, flagged: Bool = false, flagReason: String? = nil, voiceUrl: String? = nil, isSecret: Bool = false, unlockedBy: [String] = [], seenBy: [String] = []) {
        self.id = id
        self.senderId = senderId
        self.imageUrl = imageUrl
        self.timestamp = timestamp
        self.latitude = latitude
        self.longitude = longitude
        self.cityName = cityName
        self.thumbnailUrl = thumbnailUrl
        self.smallThumbnailUrl = smallThumbnailUrl
        self.flagged = flagged
        self.flagReason = flagReason
        self.voiceUrl = voiceUrl
        self.isSecret = isSecret
        self.receiverIds = receiverIds
        self.unlockedBy = unlockedBy
        self.seenBy = seenBy
    }
}

File: android/app/src/main/java/com/celalbasaran/stripmate/data/model/Strip.kt
package com.celalbasaran.stripmate.data.model

import com.google.firebase.Timestamp
import com.google.firebase.firestore.DocumentSnapshot
import java.util.Date
import java.util.UUID

data class Strip(
    val id: String = UUID.randomUUID().toString(),
    val senderId: String = "",
    val receiverIds: List<String> = emptyList(),
    val imageUrl: String = "",
    val timestamp: Date = Date(),
    val latitude: Double? = null,
    val longitude: Double? = null,
    val cityName: String? = null,
    val thumbnailUrl: String? = null,
    val smallThumbnailUrl: String? = null,
    val flagged: Boolean = false,
    val flagReason: String? = null,
    val voiceUrl: String? = null,
    val reactions: Map<String, List<String>>? = null, // emoji -> list of userIds
    val isSecret: Boolean = false,
    val unlockedBy: List<String> = emptyList(),
    val videoUrl: String? = null,
    val videoDuration: Double? = null
) {
    val isVideo: Boolean get() = videoUrl != null

    /** Bu strip gizli mi ve henüz userId tarafından açılmamış mı? */
    fun isLockedFor(userId: String): Boolean {
        if (!isSecret || senderId == userId) return false
        return !unlockedBy.contains(userId)
    }

    fun toMap(): Map<String, Any?> = buildMap {
        put("id", id)
        put("senderId", senderId)
        put("receiverIds", receiverIds)
        put("imageUrl", imageUrl)
        put("timestamp", Timestamp(timestamp))
        latitude?.let { put("latitude", it) }
        longitude?.let { put("longitude", it) }
        cityName?.let { put("cityName", it) }
        thumbnailUrl?.let { put("thumbnailUrl", it) }
        smallThumbnailUrl?.let { put("smallThumbnailUrl", it) }
        put("flagged", flagged)
        flagReason?.let { put("flagReason", it) }
        voiceUrl?.let { put("voiceUrl", it) }
        reactions?.let { put("reactions", it) }
        put("isSecret", isSecret)
        if (unlockedBy.isNotEmpty()) put("unlockedBy", unlockedBy)
        videoUrl?.let { put("videoUrl", it) }
        videoDuration?.let { put("videoDuration", it) }
    }

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromDocument(doc: DocumentSnapshot): Strip? {
            if (!doc.exists()) return null
            return Strip(
                id = doc.id,
                senderId = doc.getString("senderId") ?: "",
                receiverIds = doc.get("receiverIds") as? List<String> ?: emptyList(),
                imageUrl = doc.getString("imageUrl") ?: "",
                timestamp = doc.getTimestamp("timestamp")?.toDate() ?: Date(),
                latitude = doc.getDouble("latitude"),
                longitude = doc.getDouble("longitude"),
                cityName = doc.getString("cityName"),
                thumbnailUrl = doc.getString("thumbnailUrl"),
                smallThumbnailUrl = doc.getString("smallThumbnailUrl"),
                flagged = doc.getBoolean("flagged") ?: false,
                flagReason = doc.getString("flagReason"),
                voiceUrl = doc.getString("voiceUrl"),
                reactions = doc.get("reactions") as? Map<String, List<String>>,
                isSecret = doc.getBoolean("isSecret") ?: false,
                unlockedBy = doc.get("unlockedBy") as? List<String> ?: emptyList(),
                videoUrl = doc.getString("videoUrl"),
                videoDuration = (doc.get("videoDuration") as? Number)?.toDouble()
            )
        }
    }
}
